### ISSUE TITLE
Protect infrastructure of requirements

### ISSUE THREAD

**Issue by mluckcuck** [valu3s-mu/mu-fret#12]
The user should not be able to extract things like "shall" or "if" that are part of the 'infrastructure' of Fretish. 
The UI should prevent them by checking what's in the extract string. 

Could be done live, since there is a call back every time the field is changed.
Could be done on 'Ok'.

Could be both...

**Comment by oisinsheridan** [comment]
Brainstorming this a bit. I think that at least for some of the keywords, we should be able to take a naive approach of just checking "is this a substring of a longer name", and if not then it must be the keyword. E.g. I'm pretty sure " shall " on its own can only appear once in a requirement as that mandatory keyword, so all we need to do is check if the characters on either side of it are valid for variable names (I think?).

Unfortunately we can't just check for spaces and the end of the string; there are other characters that act as separators while retaining the keywords. E.g. In "satisfy controlObjectives" or "satisfy(controlObjectives)", 'satisfy' is still a keyword in both. It would be nice to have a list of such characters, I assume they are protected characters in NuSMV or something like that.

I think "If" is going to be a lot tougher to pull off, since "if x then y" is a valid boolean statement that the user should probably be able to refactor.

**Comment by mluckcuck** [comment]
Yes, it seems easy enough to stop a user extracting just "shall" or "if" (check if the fragment text == "shall", e.g.). 
I don't think the fragment text should be allowed to contain those keywords either, so I think we just need to check if the fragment text contains those keywords. The point you raise about "satisfy X" vs "satisfy(X)" is interesting... I suppose the character after a keyword can only be a space or "(" ... right? And before...is probably space or ")". Might be a good place to start.

Ah, I had forgotten that FRETISH could do if..then... (interestingly, the generated text and diagram explanations turn it into an implication. This one is a little more complicated and will need a regular expression or something more fancy, but it should check:
* if the fragment text contains "if" then it's invalid,
* if the fragment text contains "if ... else ..." then it's fine.

Oh, and as for the NuSMV keywords, I'd already grabbed them from the manual. They're now at the end of the README, under the "Assumptions" section.

**Comment by oisinsheridan** [comment]
> I suppose the character after a keyword can only be a space or "(" ... right? And before...is probably space or ")". 

This was exactly my thought yeah. I think that makes a lot of sense, but I was looking at the FRET grammar and I noticed this:
![image](https://github.com/valu3s-mu/mu-fret/assets/82467152/f48e91c9-17db-43a4-b738-73f4542e8bb4)

I did a bit of testing for myself and FRET will happily let you write something like "this makes shall hard to check" as a string, completely separate from the SHALL keyword. So that makes things a bit more annoying if we want to be robust

**Comment by mluckcuck** [comment]
Of course it allows that...

We could add a restriction that you can't reuse FRETISH keywords in Mu-FRETISH(?) but that feels clumsy. We may have to, though.

Maybe the robust option needs to make use of the extra information about the FRETISH that is in the requirement document in the database. In the 'developer guide', I put the data dictionary for each database. When the requirement is parsed, we get the range and text for each clause of a FRETISH requirement (component text and the range within the full text) perhaps we can use that? For example:
*  check for "shall" in the fragment text...
* if yes, check if it's within the range of any of the clauses...
* if yes, then it is fine; if no, then it's the keyword. ... maybe?
